File: pkg/platform/payment_service.go
package platform

import (
	"context"
	"errors"
)

var (
	ErrInvalidAccountData  = errors.New("invalid account data")
	ErrTransactionRejected = errors.New("transaction rejected")
)

type PaymentService interface {
	// Send sends money from one account to another.
	Send(ctx context.Context, user *User, source, des *PaymentAccount, amount int64, callbackData string) (*TransactionHistory, error)

	// Withdraw money.
	Withdraw(ctx context.Context, user *User, account *PaymentAccount, amount int64, callbackData string) (*TransactionHistory, error)

	// Get matching history from destination service/account
	GetMatchingHistory(ctx context.Context, account *PaymentAccount, history *TransactionHistory) (*TransactionHistory, error)

	// Get balance
	Balance(ctx context.Context, account *PaymentAccount) (*MonetaryAmount, error)

	// Validate user
	Validate(ctx context.Context, user *User, accountForeignID string, callbackData string) error
}

File: fastify/platform/payment_service.ts
import type { MonetaryAmount } from './monetary'
import type { PaymentAccount } from './payment_account'
import type { TransactionHistory } from './transaction_history'
import type { User } from './user'

export const ErrInvalidAccountData = new Error('invalid account data');
export const ErrTransactionRejected = new Error('transaction rejected');

export enum TransactionType {
	Withdraw,
	Send
}

export interface PaymentService {
	send(signal: AbortSignal, user: User, source: PaymentAccount, dest: PaymentAccount, amount: number, callbackData: string): Promise<TransactionHistory>;
	withdraw(signal: AbortSignal, user: User, account: PaymentAccount, amount: number, callbackData: string): Promise<TransactionHistory>;
	getMatchingHistory(signal: AbortSignal, account: PaymentAccount, history: TransactionHistory): Promise<TransactionHistory>;
	balance(signal: AbortSignal, account: PaymentAccount): Promise<MonetaryAmount>;
	validate(signal: AbortSignal, user: User, accountForeignID: string, callbackData: string): Promise<void>;
}